executor/src/witgen/affine_expression.rs

```diff
@@ -13,128 +12,116 @@ use super::{EvalError::*, EvalResult, EvalValue, IncompleteCause};
 
 /// An expression affine in the committed polynomials (or symbolic variables in general).
 #[derive(Debug, Clone)]
-pub struct AffineExpression<K, T> {
-    pub coefficients: Vec<(K, T)>,
-    pub offset: T,
-    /// If true, all coefficients have nonzero values and there are no duplicate keys
-    /// (a constant affine expression is always clean).
-    clean: bool,
+pub enum AffineExpression<K, T> {
+    Constant(T),
+    One((K, T), T),
```

Comment:
Maybe `OneVar`?

Reply:
Did you try using `Option<T>` for the second `T` (optimizing the case where it is zero)?

Reply:
you mean wrapping the offset in Option<>? If so, I think it wouldn't change much and complicate all the `match`es in the code